Split BadgeRewarder into multiple services

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Refactor

## Description

This PR removes `BadgeRewarder` from `app/labor` and splits it into multiple services in `app/services`.

Some general notes:

1. A central piece of core logic was factored out into a `Badges::Award` service that the other services call.
2. A lot of the code tried to save the user after a badge was awarded. I changed this to use `touch` instead, which I think is what we actually want.
3. I made some minor cleanups along the way, e.g. rewriting queries for the `ActiveRecord` DSL.
4. The PR looks bigger than it actually is since I primarily just moved code from methods to separate services.

## Related Tickets & Documents

https://github.com/forem/InternalProjectPlanning/issues/271

## QA Instructions, Screenshots, Recordings

1. Specs should be green
2. You can try manually running the badges rake tasks to assign badges.

## Added tests?

- [X] Yes: moved, updated

## Added to documentation?

- [X] No documentation needed

## [optional] Are there any post deployment tasks we need to perform?

Maybe some scheduler updates, pending confirmation from @maestromac.

citizen428 commented on app/workers/badge_achievements/badge_award_worker.rb:
```diff
@@ -5,10 +5,14 @@ class BadgeAwardWorker
     sidekiq_options queue: :high_priority, retry: 10
 
     def perform(usernames, badge_slug, message)
```
@mac Can you please help me verify how we award badges? Is most of it done through Heroku scheduler?

1. The first part of this `if` will take care of things that don't need extra arguments (yearly club, tags, 4, 8, and 16 week streak).
2. The second part here directly awards badges, what are we using this for?
3. The Rake tasks awards top seven, fab five, and contributor, how are these scheduled?
4. What about` award_contributor_badges_from_github` (now `Badges::AwardContributorFromGithub`)? This doesn't seem to be covered by any of these cases, are we still using it?

maestromac replied:
@citizen428 Sorry I missed this. We aren't using Heroku Scheduler anymore. Instead, we are using `sidekiq-cron` gem (shoutout to @mstruve !). You can see each individual task schedule in `config/schedule.yml`.
2. I'm not sure quite sure either. I believe it for one-off awards. cc @benhalpern
3. & 4. You can see each individual task schedule in `config/schedule.yml`.

citizen428 replied:
Thanks @maestromac. I had actually looked for a file like this, no idea how I missed it 🤷

citizen428 commented:
I updated the schedule now thanks @maestromac!

rhymes commented on app/services/badges/award_yearly_club.rb:
```diff
@@ -0,0 +1,35 @@
+    private
+
+    def generate_message(years)
+      "Happy #{SiteConfig.community_name} birthday! " \
+        "Can you believe it's been #{years} #{'year'.pluralize(years)} already?!"
```
same idea about freezing the template

citizen428 replied:
✅

rhymes commented on config/schedule.yml:
```diff
@@ -21,49 +21,49 @@ award_yearly_club_badges:
   class: "BadgeAchievements::BadgeAwardWorker"
```
**potentially blocking**: will this break if there are serialized workers with the old argument after the restart deploy?

citizen428 replied:
Most of them run only every 12 hours, one only runs weekly. I don't think there should be too many queued up jobs hanging around. The only exception is `award_contributor_badges_from_github` which runs hourly at 20 minutes past the full hour, but even there we could probably miss one in the worst case and just wait for the next run to award the badges.

rhymes replied:
Sure! We can also re-run the manually if we need to

rhymes reviewed (commented):
What a refactoring! Left a potentially blocking note on `schedule.yml` and a bunch of other notes

citizen428 commented:
Thanks for all the feedback @rhymes. I'll have a look tomorrow, so far I mostly focused on the splitting and only did some very minor cleanups along the way.

maestromac commented:
@citizen428 yeah it looks like a manual rake task in `lib/tasks/badges.rake`

maestromac commented on app/workers/badge_achievements/badge_award_worker.rb:
```diff
@@ -5,10 +5,14 @@ class BadgeAwardWorker
     def perform(usernames, badge_slug, message)
-      if BadgeRewarder.respond_to?(badge_slug)
-        BadgeRewarder.public_send(badge_slug)
+      if (award_class = "Badges::#{badge_slug.classify}".safe_constantize)
+        award_class.call
```
Love it :star_struck:

maestromac reviewed (approved):
Love the refactor! LGTM!

citizen428 commented on app/services/badges/award.rb:
```diff
@@ -0,0 +1,16 @@
+module Badges
+  class Award
+    def self.call(user_relation, slug, message_markdown)
+      badge_id = Badge.find_by(slug: slug)&.id
+      return unless badge_id
```
✅

citizen428 commented on app/services/badges/award_beloved_comment.rb:
```diff
@@ -0,0 +1,36 @@
+    end
+
+    def call
+      badge_id = Badge.find_by(slug: BADGE_SLUG)&.id
```
✅

citizen428 commented on app/services/badges/award_streak.rb:
```diff
@@ -0,0 +1,58 @@
+    end
+
+    def call
+      badge_id = Badge.find_by(slug: "#{weeks}-week-streak")&.id
```
✅

citizen428 commented on app/services/badges/award_beloved_comment.rb:
```diff
@@ -0,0 +1,36 @@
+    attr_reader :comment_count
+
+    def generate_message(comment)
+      "You're famous! " \
+        "[This is the comment](#{URL.comment(comment)}) for which you are being recognized. 😄"
```
✅

citizen428 commented on app/services/badges/award_streak.rb:
```diff
@@ -0,0 +1,58 @@
+    def generate_message
+      return LONGEST_STREAK_MESSAGE if weeks == LONGEST_STREAK_WEEKS
+
+      "Congrats on achieving this streak! Consistent writing is hard. " \
+        "The next streak badge you can get is the #{weeks * 2} Week Badge. 😉"
```
✅

citizen428 commented on app/services/badges/award_eight_week_streak.rb:
```diff
@@ -0,0 +1,7 @@
+module Badges
+  class AwardEightWeekStreak
+    def self.call
+      ::Badges::AwardStreak.call(8)
```
Yes, indeed, I made this change now.

citizen428 commented on app/services/badges/award_contributor_from_github.rb:
```diff
@@ -0,0 +1,28 @@
+    REPOSITORIES = [
+      "forem/forem",
+      "forem/forem-browser-extension",
+      "thepracticaldev/DEV-Android",
+      "thepracticaldev/DEV-ios",
```
✅

citizen428 commented on app/services/badges/award_fab_five.rb:
```diff
@@ -0,0 +1,13 @@
+  class AwardFabFive
+    BADGE_SLUG = "fab-5".freeze
+
+    def self.call(usernames, message_markdown = "Congrats!!!")
```
✅

citizen428 commented on app/services/badges/award_top_seven.rb:
```diff
@@ -0,0 +1,13 @@
+  class AwardTopSeven
+    BADGE_SLUG = "top-7".freeze
+
+    def self.call(usernames, message_markdown = "Congrats!!!")
```
✅

citizen428 commented on app/services/badges/award_tag.rb:
```diff
@@ -0,0 +1,36 @@
+      "Congratulations on posting the most beloved [##{tag.name}](#{URL.tag(tag)}) post " \
+        "from the past seven days! " \
+        "Your winning post was [#{winning_article.title}](#{URL.article(winning_article)}). " \
+        "(You can only win once per badge-eligible tag)"
```
✅

citizen428 commented:
@rhymes I addressed all of your feedback now.

@maestromac Just making sure you didn't miss [this discussion](https://github.com/forem/forem/pull/11902#discussion_r547212497), as it's potentially relevant for SRE.

rhymes reviewed (approved):
LGTM!

maestromac commented:
@citizen428 Yep I saw it. It shouldn't be a concern.
